code/modules/antagonists/hivemind/hivemind.dm

```diff
@@ -0,0 +1,194 @@
+	antag_moodlet = /datum/mood_event/focused
+	var/special_role = ROLE_HIVE
+	var/list/hivemembers = list()
+	var/hive_size = 0
+
+	var/list/upgrade_tiers = list(
+		//Tier 1
+		/obj/effect/proc_holder/spell/target_hive/hive_add = 0,
+		/obj/effect/proc_holder/spell/target_hive/hive_remove = 0,
+		/obj/effect/proc_holder/spell/target_hive/hive_see = 0,
+		/obj/effect/proc_holder/spell/target_hive/hive_shock = 2,
+		/obj/effect/proc_holder/spell/self/hive_drain = 4,
+		//Tier 2
+		/obj/effect/proc_holder/spell/target_hive/hive_warp = 6,
+		/obj/effect/proc_holder/spell/targeted/hive_hack = 8,
+		/obj/effect/proc_holder/spell/target_hive/hive_control = 10,
+		/obj/effect/proc_holder/spell/targeted/induce_panic = 12,
+		//Tier 3
+		/obj/effect/proc_holder/spell/self/hive_loyal = 15,
+		/obj/effect/proc_holder/spell/targeted/hive_assim = 18,
+		/obj/effect/proc_holder/spell/targeted/forcewall/hive = 20,
+		/obj/effect/proc_holder/spell/target_hive/hive_attack = 25)
+
+/datum/antagonist/hivemind/proc/calc_size()
+	listclearnulls(hivemembers)
+	var/old_size = hive_size
+	hive_size = hivemembers.len
+	if(hive_size != old_size)
+		check_powers()
+
+/datum/antagonist/hivemind/proc/check_powers()
+	for(var/power in upgrade_tiers)
+		var/level = upgrade_tiers[power]
+		if(hive_size >= level && !(locate(power) in owner.spell_list))
+			owner.AddSpell(new power(null))
+		else if(hive_size < level && (locate(power) in owner.spell_list))
+			owner.RemoveSpell(power)
+
+
+/datum/antagonist/hivemind/proc/add_to_hive(var/mob/living/carbon/human/H)
+	var/warning = "<span class='userdanger'>We detect a surge of psionic energy from [H.real_name] before they disappear from the hive. An enemy host, or simply a stolen vessel?</span>"
+	var/user_warning = "<span class='userdanger'>We have detected an enemy hivemind using our physical form as a vessel and have begun ejecting their mind! They will be alerted of our disappearance once we succeed!</span>"
+	for(var/datum/antagonist/hivemind/enemy_hive in GLOB.antagonists)
+		if(H.mind == enemy_hive.owner)
+			var/eject_time = rand(1400,1600) //2.5 minutes +- 10 seconds
+			addtimer(CALLBACK(GLOBAL_PROC, /proc/to_chat, H, user_warning), rand(500,1300)) // If the host has assimilated an enemy hive host, alert the enemy before booting them from the hive after a short while
+			addtimer(CALLBACK(GLOBAL_PROC, /proc/to_chat, owner, warning), eject_time) //As well as the host who just added them as soon as they're ejected
+			addtimer(CALLBACK(GLOBAL_PROC, /proc/remove_hivemember, H), eject_time)
+	hivemembers |= H
+	calc_size()
+
+/datum/antagonist/hivemind/proc/remove_from_hive(var/mob/living/carbon/human/H)
+	hivemembers -= H
+	calc_size()
+
+/datum/antagonist/hivemind/proc/destroy_hive()
+	hivemembers = list()
+	calc_size()
+
+/datum/antagonist/hivemind/antag_panel_data()
+	return "Vessels Assimilated: [hive_size]"
+
+/datum/antagonist/hivemind/on_gain()
+
+	owner.special_role = special_role
+	apply_innate_effects()
```

Comment:
This will be done in parent call already, if order is important extract that. But from what i can see it doesn't matter.

Reply:
Removed